### PR TITLE
Enhance mnist training test.

### PR DESCRIPTION
**Pull request by cthsieh**
### Ticket
None

### Problem description
After reviewing `test_mnist.py` as an example of testing the training mode, I noticed that it lacks a test for the backward pass, which calculates the gradients. I've rewritten the test, and if you agree with this change, I plan to apply the same approach to all other training tests.

### What's changed
* The training test in `test_mnist.py`.
* The training process in `conftest.py`.

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -83,6 +83,25 @@ def compile_and_run(device, reset_torch_dynamo, request):
 
     if "torch_ttnn" in record:
         model, inputs, outputs = record["torch_ttnn"]
+
+        try:
```
Comment (ayerofieiev-tt) [inline tests/conftest.py:87]: I am not clear why you need this try/catch block, it does not call any functions and I can’t see how it can throw.

Comment (ayerofieiev-tt) [reply tests/conftest.py:87]: Is this for the case when inputs or model do not contain a key?

File: tests/models/mnist/test_mnist.py [hunk 2]
```diff
@@ -37,6 +37,14 @@ def forward(self, x):
 def test_mnist_train(record_property):
     record_property("model_name", "Mnist (Train)")
 
+    # Fixing the random seed for reproducibility to ease debugging.
+    #
+    # Training processes involve more randomness compared to evaluation,
+    # such as random initialization of weights.
+    # Setting a fixed random seed is crucial for consistent testing
+    # and debugging during the training process.
+    torch.manual_seed(99)
```
Comment (ayerofieiev-tt) [inline tests/models/mnist/test_mnist.py:46]: 👍

File: tests/conftest.py (hunk 1 above)
Comment (kevinwuTT) [reply tests/conftest.py:87]: Yeah, at least `requires_grad` is an attribute of `torch.Tensor` [type](https://pytorch.org/docs/stable/generated/torch.Tensor.requires_grad_.html). I've seen other models have other types for their inputs (see below), so I think this PR only supports `Tensor` input types at the moment. 

https://github.com/tenstorrent/pytorch2.0_ttnn/blob/6f78bd3a6497a8bb5246e3d46fce431cadc50c37/tests/conftest.py#L172-L178

**Review by ayerofieiev-tt — APPROVED** [review]

**Comment by kevinwuTT** [comment]
I think this looks good. Just a note, we will need to support the new backwards ops and most of them will return a tuple of outputs.

Screenshot 2024-09-23 at 11 13 23 AM

**Review by kevinwuTT — APPROVED** [review]

File: tests/conftest.py (hunk 1 above)
Comment (cthsieh) [reply tests/conftest.py:87]: The try/catch is to handle the fact that `inputs.requires_grad` is invalid if `inputs` is an instance of type `collections.Mapping` or `collections.Sequence`.

In other words, we assume that the models under train test must have only one input.

Will relax the assumption if necessary.
